src/LuaColor.cpp

```diff
@@ -0,0 +1,213 @@
+{
+	r = std::max(std::min(r, float(255.0)), float(0.0));
+	g = std::max(std::min(g, float(255.0)), float(0.0));
+	b = std::max(std::min(b, float(255.0)), float(0.0));
+	a = std::max(std::min(a, float(255.0)), float(0.0));
+}
+
+static int l_color_new(lua_State *L)
+{
+	LUA_DEBUG_START(L);
+	double r = luaL_checknumber(L, 1);
+	double g = luaL_checknumber(L, 2);
+	double b = luaL_checknumber(L, 3);
+	double a;
+	if (lua_isnumber(L, 4)) a = luaL_checknumber(L, 4);
+	else a = 255.0;
+	sanity_check(r, g, b, a);
+	LuaColor::PushToLua(L, Color(r, g, b, a));
+	LUA_DEBUG_END(L, 1);
+	return 1;
+}
+
+static int l_color_call(lua_State *L)
+{
+	LUA_DEBUG_START(L);
+	double r = luaL_checknumber(L, 2);
+	double g = luaL_checknumber(L, 3);
+	double b = luaL_checknumber(L, 4);
+	double a;
+	if (lua_isnumber(L, 5)) a = luaL_checknumber(L, 5);
+	else a = 255.0;
+	sanity_check(r, g, b, a);
+	Color c = Color(r, g, b, a);
+	LuaColor::PushToLua(L, Color(r, g, b, a));
+	LUA_DEBUG_END(L, 1);
+	return 1;
+}
+
+static int l_color_tostring(lua_State *L)
+{
+	const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+	luaL_Buffer buf;
+	luaL_buffinit(L, &buf);
+	char *bufstr = luaL_prepbuffer(&buf);
+	int len = snprintf(bufstr, LUAL_BUFFERSIZE, "Color(%i, %i, %i, %i)", c->r, c->g, c->b, c->a);
+	assert(len < LUAL_BUFFERSIZE); // XXX should handle this condition more gracefully
+	luaL_addsize(&buf, len);
+	luaL_pushresult(&buf);
+	return 1;
+}
+
+static int l_color_add(lua_State *L)
+{
+	const Color4ub *a = LuaColor::CheckFromLua(L, 1);
+	const Color4ub *b = LuaColor::CheckFromLua(L, 2);
+	LuaColor::PushToLua(L, *a + *b);
+	return 1;
+}
+
+static int l_color_mul(lua_State *L)
+{
+	if (lua_isnumber(L, 1)) {
+		const double s = lua_tonumber(L, 1);
+		const Color4ub *c = LuaColor::CheckFromLua(L, 2);
+		LuaColor::PushToLua(L, *c * s);
+	} else if (lua_isnumber(L, 2)) {
+		const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+		const double s = lua_tonumber(L, 2);
+		LuaColor::PushToLua(L, *c * s);
+	} else {
+		return luaL_error(L, "Cannot multiply a Color time Color");
+	}
+	return 1;
+}
+
+static int l_color_new_index(lua_State *L)
+{
+	Color4ub *c = LuaColor::CheckFromLua(L, 1);
+	if (lua_type(L, 2) == LUA_TSTRING) {
+		const char *attr = luaL_checkstring(L, 2);
+		if (!strcmp(attr, "r")) {
+			c->r = luaL_checknumber(L, 3);
+		} else if (!strcmp(attr, "g")) {
+			c->g = luaL_checknumber(L, 3);
+		} else if (!strcmp(attr, "b")) {
+			c->b = luaL_checknumber(L, 3);
+		} else if (!strcmp(attr, "a")) {
+			c->a = luaL_checknumber(L, 3);
+		} else {
+			luaL_error(L, "Index '%s' is not available: use 'r', 'g', 'b' or 'a'", attr);
+		}
+
+	} else {
+		luaL_error(L, "Expected Color, but type is '%s'", lua_typename(L, lua_type(L, 2)));
+	}
+	LuaColor::PushToLua(L, *c);
+	return 1;
+}
+
+static int l_color_index(lua_State *L)
+{
+	const Color4ub *c = LuaColor::CheckFromLua(L, 1);
+	if (lua_type(L, 2) == LUA_TSTRING) {
+		const char *attr = luaL_checkstring(L, 2);
```

Comment:
This is redundant as you've already checked the type of the value - there's no possible way this will error out. Additionally, you should check that the length of the string = 1, and gate on that as well.

Reply:
> This is redundant as you've already checked the type of the value

Copy-paste error: `LuaVector` has an `else` branch...

>Additionally, you should check that the length of the string = 1, and gate on that as well.

I don't think there's needs for this.

Reply:
> I don't think there's needs for this.

It saves four invocations of `strcmp` every time someone calls `Shade` or `Tint`, which is going to happen at least once per frame(and likely tens or hundreds of times). Actually, use `lua_tolstring(l, const char *str, size_t *len)` and just check `len` to see if it's a single-character string - then you don't even need to `strcmp`, you can just check `attr[0] = 'r'` directly, saving function-call overhead.

Reply:
How you can call this method with a zero-length string, given Lua parser?

Reply:
...I'm not sure if you're talking about `attr[0]` or checking that it's a single-character string; arrays in C are zero-indexed, and the single-character check is specifically to speed up method lookups for functions like Shade or Tint. We're not concerned with zero-length strings but with strings > 1.

Reply:
Specifically, the reason we check the string length by calling `lua_tolstring` instead of `luaL_checkstring` is because as-written, calling code can use `Color().redABCDDESLDGS` and the current `__index` method will think that you've written `Color().r`. If you use the single-character comparison, you _must_ check the string length.

Reply:
> you must check the string length.

I don't think so: I can show you a lot of examples here and there on code in which you can ends with CTD or strange behaviour because of this kind of errors not handled so well in C.
...And not in a "critical code" like this, called _a lot_ each frame.

Thus I'm wondering if there are reasons for you to ask for so many changes:
I changed what you asked for inside reasonable limits (...color "swizzling"?) l but
I begin to suspect there may be something else you want to say.

Let me know...

Reply:
Okay, look. I'm not trying to badger you for unreasonable demands, just trying to help you write better, less-error-prone code. Here's an example of that function rewritten to catch the most useful edge cases:

```C++
const Color4ub *c = LuaColor::CheckFromLua(L, 1);
size_t len = 0;
const char *attr = NULL;
if (lua_type(L, 2) == LUA_TSTRING)
    attr = lua_tolstring(L, 2, &len);
if (attr != NULL && len == 1) {
    if (attr[0] == 'x') {
        ...
```

This does two things - first of all, it reduces the overhead involved with calling `luaL_checkstring`, and secondly it correctly handles cases where you might _intend_ to be looking up a method called e.g. `redify`, and do not want to get the red channel of the color instead.

As an aside, give [the wiki article on swizzling](https://en.wikipedia.org/wiki/Swizzling_(computer_graphics)) a read.